controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +58,787 @@ ControllerManager::load_controller(
     [&](auto loader)
     {return loader->is_available(controller_type);});
 
-  std::shared_ptr<controller_interface::ControllerInterface> controller(nullptr);
+  controller_interface::ControllerInterfaceSharedPtr controller(nullptr);
   if (it != loaders_.cend()) {
     controller = (*it)->create(controller_type);
   } else {
     const std::string error_msg("Loader for controller '" + controller_name + "' not found\n");
     RCLCPP_ERROR(get_logger(), "%s", error_msg.c_str());
     throw std::runtime_error(error_msg);
   }
+  ControllerSpec controller_spec;
+  controller_spec.c = controller;
+  controller_spec.info.name = controller_name;
+  controller_spec.info.type = controller_type;
 
-  return add_controller_impl(controller, controller_name);
+  return add_controller_impl(controller_spec);
 }
 
-std::vector<std::shared_ptr<controller_interface::ControllerInterface>>
-ControllerManager::get_loaded_controllers() const
+controller_interface::ControllerInterfaceSharedPtr ControllerManager::load_controller(
+  const std::string & controller_name)
 {
-  return loaded_controllers_;
+  const std::string param_name = controller_name + ".type";
+  std::string controller_type;
+
+  // A priori we don't know the name of the controllers that will be loaded, so we cannot
```

Comment:
What I wanted to express was that we cannot declare in the controller_manager code the name of the parameters of the controllers that will be loaded in the future, because they are plugins and we cannot be aware of all of them.

Reply:
It's set at the controller_manager node.

Let's say you want to load a controller called "left_leg_controller" of type "LegController".

Right now at the controller manager node you set a parameter called `left_leg_controller.type` with value `"LegController"`. I would change this and put these parameters into a `controllers` field or something.

So when someone does a service call load_controller("left_leg_controller"), you are able to read that param and instantiate the proper controller type. 

Probably the params for controller_manager node will look like:
```
controller_manager:
  controllers:
    left_leg_controller: "LegController"
```

And for the left_leg_controller itself:
```
left_leg_controller:
  joint_names: ["left_leg_1_joint", "left_leg_1_joint"]
  another_controller_param: False
```

I was basing myself on [the costmaps config](https://github.com/ROBOTIS-GIT/turtlebot3/blob/foxy-devel/turtlebot3_navigation2/param/waffle.yaml#L152) but there, the costmap layers are not nodes.

Would this structure address your concerns? Also keep in mind that we should support multiple controllers of the same type with different names and configurations (ie force torque sensor controllers).

Reply:
I guess I was just wondering what's the use case of loading a controller exclusively by name compared to loading it by name and type. The reason why I am bringing this up is that it feels a bit like we disable the modularity of dynamically loading controllers, as we have to first populate the type of the controller to the controller manager and then later on load the controller itself. I was just under the impression that the controller manager should be basically agnostic of the type of the controllers.

Reply:
In ROS1 it made sense because you could have in a single namespace the type + other params of a controller. 
And with the name of the controller you could load, start and stop.
For instance: https://github.com/pal-robotics/tiago_robot/blob/kinetic-devel/tiago_controller_configuration/config/current_limit_controllers.yaml

In ROS2, if they have to be parameters of separate nodes, that may not make sense. 
Your thoughts @bmagyar ?

Reply:
This is a tricky situation. I'd bring up another option that doesn't seem particularly inviting either:

load all controller parameters into the controller_manager namespace and hand the appropriate section over to the node at load time. It's a little hacky but would achieve something similar to what we are looking for.